src/merkle_tree/mod.rs

```diff
@@ -234,26 +239,145 @@ impl<P: Config> MerkleTree<P> {
+    }
+
+    #[cfg(feature = "parallel")]
+    pub fn new_with_leaf_digest(
+        leaf_hash_param: &LeafParam<P>,
+        two_to_one_hash_param: &TwoToOneParam<P>,
+        leaves_digest: Vec<P::LeafDigest>,
+    ) -> Result<Self, crate::Error> {
+        // Parallel implementation of new_with_leaf_digest
+
+        let leaf_nodes_size = leaves_digest.len();
+        assert!(
+            leaf_nodes_size.is_power_of_two() && leaf_nodes_size > 1,
+            "`leaves.len() should be power of two and greater than one"
+        );
+        let non_leaf_nodes_size = leaf_nodes_size - 1;
+
+        let tree_height = tree_height(leaf_nodes_size);
+
+        let hash_of_empty: P::InnerDigest = P::InnerDigest::default();
+
+        // initialize the merkle tree as array of nodes in level order
+        let mut non_leaf_nodes: Vec<P::InnerDigest> = (0..non_leaf_nodes_size)
+            .into_par_iter()
+            .map(|_| hash_of_empty.clone())
+            .collect();
+
+        // Compute the starting indices for each non-leaf level of the tree
+        let mut index = 0;
+        let mut level_indices = Vec::with_capacity(tree_height - 1);
+        for _ in 0..(tree_height - 1) {
+            level_indices.push(index);
+            index = left_child(index);
+        }
 
-        // compute and store hash values for each leaf
-        for leaf in leaves.into_iter() {
-            leaves_digests.push(P::LeafHash::evaluate(leaf_hash_param, leaf)?)
+        // compute the hash values for the non-leaf bottom layer
+        {
+            let start_index = level_indices.pop().unwrap();
+            let upper_bound = left_child(start_index);
+            let bottom_nodes = &mut non_leaf_nodes[start_index..upper_bound];
+
+            bottom_nodes.par_iter_mut().enumerate().for_each(|(i, n)| {
+                // `left_child(current_index)` and `right_child(current_index) returns the position of
+                // leaf in the whole tree (represented as a list in level order). We need to shift it
+                // by `-upper_bound` to get the index in `leaf_nodes` list.
+
+                //similarly, we need to rescale i by start_index
+                //to get the index outside the slice and in the level-ordered list of nodes
+
+                let current_index = i + start_index;
+                let left_leaf_index = left_child(current_index) - upper_bound;
+                let right_leaf_index = right_child(current_index) - upper_bound;
+                // compute hash
+                *n = P::TwoToOneHash::evaluate(
+                    two_to_one_hash_param,
+                    P::LeafInnerDigestConverter::convert(leaves_digest[left_leaf_index].clone())
+                        .unwrap(),
+                    P::LeafInnerDigestConverter::convert(leaves_digest[right_leaf_index].clone())
+                        .unwrap(),
+                )
+                .unwrap()
```

Comment:
and here

Reply:
Unfortunately we cannot without introducing some changes to the API, refer to @WizardOfMenlo comment above https://github.com/arkworks-rs/crypto-primitives/pull/125#issuecomment-1867841425